spec/02-integration/04-admin_api/21-client_certificate.lua

```diff
@@ -0,0 +1,304 @@
+            json = cjson.decode(body)
+            assert.equal(0, #json.data)
+
+            res = client:post("/services",{
+                body = {
+                    host = 'test.com',
+                    name = 'nginx',
+                    protocol = 'https',
+                    port = 443,
+                    client_certificate = {id = cert_id}
+                },
+                headers = { ["Content-Type"] = "application/json" }
+            })
+            assert.res_status(201, res)
+
+
+            -- check service
+            res = client:get("/services")
+            body = assert.res_status(200, res)
+            json = cjson.decode(body)
+            assert.equal(1, #json.data)
+            srv_id = json.data[1].id
+
+
+            -- create route
+            res = client:get("/routes")
+            body = assert.res_status(200, res)
+            json = cjson.decode(body)
+            assert.equal(0, #json.data)
+
+            res = client:post("/routes",{
+                body = {
+                    name = 'nginx',
+                    paths = {'/test'},
+                    service = {id = srv_id}
+                },
+                headers = { ["Content-Type"] = "application/json" }
+            })
+            assert.res_status(201, res)
+
+            -- check route
+            res = client:get("/routes")
+            body = assert.res_status(200, res)
+            json = cjson.decode(body)
+            assert.equal(1, #json.data)
+            route_id = json.data[1].id
+        end)
+
+        lazy_teardown(function()
+            if client then client:close() end
+            helpers.stop_kong()
+        end)
+
+        it("Send request", function ()
+            local client = helpers.http_client("localhost", 9000)
+            local res = client:get("/test")
+            res_body = assert.res_status(400, res)
+        end)
+
+        it("Change cert #Change", function()
+            -- change cert
+            local res = client:patch("/certificates/" .. cert_id, {
+                body = {
+                    cert = new_cert,
+                    key = new_key,
+                },
+                headers = { ["Content-Type"] = "application/json" }
+            })
+            local body = assert.res_status(200, res)
+
+            -- check cert
+            res  = client:get("/certificates")
+            body = assert.res_status(200, res)
+            json = cjson.decode(body)
+            assert.equal(1, #json.data)
+            assert.equals(json.data[1].cert, new_cert)
+            assert.equals(json.data[1].key, new_key)
+            cert_id = json.data[1].id
+        end)
+
+        it("Send request after change", function ()
+            local client = helpers.http_client("localhost", 9000)
+            local res = client:get("/test")
+            local body = assert.res_status(400, res)
+            print(body)
+            assert.not_equals(body, res_body)
+        end)
+    end)
+end
```

Comment:
nitpick: end of new line
```suggestion
end

```